src/NHibernate/Linq/Visitors/QueryModelVisitor.cs

```diff
@@ -551,6 +556,46 @@ private void AddJoin(JoinClause joinClause, QueryModel queryModel, bool innerJoi
 			_hqlTree.AddFromClause(join);
 		}
 
+		private class BaseMemberChecker : NhExpressionVisitor
+		{
+			private readonly ISessionFactoryImplementor _sessionFactory;
+			private bool _result;
+
+			public BaseMemberChecker(ISessionFactoryImplementor sessionFactory)
+			{
+				_sessionFactory = sessionFactory;
+			}
+
+			public bool ContainsBaseMember(Expression node)
+			{
+				_result = false;
+				Visit(node);
+
+				return _result;
+			}
+
+			protected override Expression VisitMember(MemberExpression node)
+			{
+				if (ExpressionsHelper.TryGetMappedType(
+						_sessionFactory,
+						node,
+						out _,
+						out var persister,
+						out _,
+						out var propertyPath) &&
+					persister is IOuterJoinLoadable joinLoadable &&
```

Comment:
Maybe skip property checks if no subclasses? Something like
```C#
persister.EntityMetamodel.HasSubclasses &&
persister is IOuterJoinLoadable joinLoadable ...
```

Or skip this visitor completely if `joinClause` entity has no subclasses?

Reply:
> Or skip this visitor completely if joinClause entity has no subclasses?

Added a check before visiting the node.